### PR TITLE
[Spec][DSpark] Clamp per-request verify window to context/generation budget (fix #33454)

### LINKED ISSUE #33454: [Bug] DSpark verify window crosses the model context boundary and causes an illegal RoPE read
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Environment info and a minimal source-level reproduction are included.
- [x] This is a bug report, not a general question.
- [x] This issue is written in English.

### Describe the bug

DSpark always allocates a full verification window, even when a request has fewer valid context positions remaining. At the context boundary this creates positions greater than or equal to `max_position_embeddings` and can crash the target model with an illegal RoPE table read.

Observed with DeepSeek-V4-Flash, `context_len=1048576`, DSpark `gamma=5`, and `verify_num_draft_tokens=6`. The request dump contained one request with 145,747 input tokens and 902,822 output tokens. The server then reached `full token: 1048576` and TP1 crashed.

CUDA core:

```text
CUDA Exception: Warp Illegal Address
Kernel: fused_q_norm_rope<__nv_bfloat16, 512, 64, long, true>
Grid: (40,1,1)
Block: (128,1,1)
Fault block/thread: block 16, thread 64
Fault instruction: LDG.E.64 R2,desc[UR8][R2.64]
```

The failed instruction is the 64-bit load of one RoPE frequency pair from `freqs_cis[position]`. TP0 was waiting in custom all-reduce after TP1 failed; all-reduce was not the initial fault. Both GPUs reported zero volatile corrected and uncorrected ECC errors.

The same boundary bug is still present on current main (`572924634b9f18ea86da039929db4337508c89af`):

- [`build_block_pos_offsets()`](https://github.com/sgl-project/sglang/blob/572924634b9f18ea86da039929db4337508c89af/python/sglang/srt/speculative/draft_worker_common.py#L214-L215) always returns `arange(verify_num_draft_tokens)`.
- [`uniform_ragged_layout()`](https://github.com/sgl-project/sglang/blob/572924634b9f18ea86da039929db4337508c89af/python/sglang/srt/speculative/dspark_components/dspark_planner.py#L737-L761) assigns the full verify length to every request.
- [`alloc_verify_window()`](https://github.com/sgl-project/sglang/blob/572924634b9f18ea86da039929db4337508c89af/python/sglang/srt/speculative/dspark_components/dspark_planner.py#L843-L866) computes every position as `seq_len + block_pos_offset` without checking the remaining request or context budget.

Expected behavior: clamp each request's DSpark verify length to its remaining generation and model-context budget. Every position passed to the model must be smaller than `max_position_embeddings`.

Related: #30570 reports a different context-boundary failure in EAGLE/DSA. This report is for the DSpark verify planner and the resulting RoPE illegal address.

### Reproduction

Server:

```bash
python3 -m sglang.launch_server \
  --model-path DeepSeek-V4-Flash-0731 \
  --tp 2 \
  --speculative-algorithm DSPARK \
  --moe-runner-backend flashinfer_mxfp4 \
  --chunked-prefill-size 8192
```

Send a request without an explicit `max_new_tokens` and let it decode to the model context boundary.

The position overflow can also be shown directly with the deployed DSpark helper:

```python
import torch
from sglang.srt.speculative.draft_worker_common import build_block_pos_offsets

context_len = 1_048_576
seq_lens = torch.tensor([context_len - 2])
offsets = build_block_pos_offsets(length=6, device=torch.device("cpu"))
print((seq_lens[:, None] + offsets).tolist())
```

Observed:

```text
[[1048574, 1048575, 1048576, 1048577, 1048578, 1048579]]
```

Positions `1048576` through `1048579` are outside a RoPE table with 1,048,576 rows.

### Environment

```text
SGLang source: fdebc938f7f4d16fe6b9f55dcd9a767cf0899ea1
SGLang: 0.5.16
sglang-kernel: 0.4.5
Image: lmsysorg/sglang:v0.5.16
Python: 3.12.3
GPU: 2x NVIDIA B300 SXM6 AC, compute capability 10.3
CUDA: 13.0
CUDA driver: 580.126.20
PyTorch: 2.11.0+cu130
Triton: 3.6.0
Transformers: 5.12.1
flashinfer: 0.6.14
```

### PR DESCRIPTION
**Pull request by haimianxing**
## Motivation

Fixes #33454.

DSpark allocates a full verification window for every request regardless of
remaining model-context or generation budget.  Near the 1M context boundary
this produces positions `>= max_position_embeddings`, and the RoPE kernel
crashes with `CUDA Exception: Warp Illegal Address` on the 64-bit load from
`freqs_cis[position]`.

The three root-cause sites are unchanged on current main:
- `build_block_pos_offsets()` always returns `arange(verify_num_draft_tokens)`
- `uniform_ragged_layout()` assigns the full verify length to every request
- `alloc_verify_window()` computes `seq_len + block_pos_offset` without a budget check

## Modifications

**`draft_worker_common.py`** — new `clamp_verify_lens` helper:
```python
actual = min(requested_verify_len, context_remaining, generation_remaining)
```
where `context_remaining = max_position_embeddings - seq_len` (clamped ≥ 0).

**`dspark_worker_v2.py`**:
- Per-request `remaining_generation_tokens` computed at the top of `target_step`.
- `max_new_tokens=None` guard: the original repro ("no explicit `max_new_tokens`") passed `None` to the subtraction, causing a `TypeError` before the RoPE crash. `None` is now treated as no generation cap.
- After `schedule_layout`, intersects planner layout with clamped lens. If any request is clipped, constructs a `RaggedVerifyLayout` so only legal positions reach the model. `run_compact=True` whenever a layout exists, so `run_non_compact` is never reached for a clipped batch.
- Comments document the single-thread scheduler guarantee (race-free budget snapshot) and TP consistency invariant (`batch.seq_lens` is broadcast-synced across ranks before this call).

**Note on Bug B (#33579 / PR #33581):** A second independent bug exists at `page_size=1` (silent KV slot leak). That is orthogonal to this PR and addressed in #33581. `test_allocation_sizing.py` documents both the current failure and expected fixed behaviour as a regression fence.

## Verification

All tests are CPU-only (no GPU required). 12 passed in 1.53s:

```
test/registered/unit/spec/test_dspark_context_boundary.py   6 passed
test/registered/unit/mem_cache/test_allocation_sizing.py    6 passed
```

Boundary trace (`max_position_embeddings=100`):
```
request=0 seq_len=10  context_remaining=90 generation_remaining=20 → verify_len=6
request=1 seq_len=98  context_remaining=2  generation_remaining=20 → verify_len=2
request=2 seq_len=99  context_remaining=1  generation_remaining=20 → verify_len=1
request=3 seq_len=100 context_remaining=0  generation_remaining=20 → verify_len=0
request=4 seq_len=97  context_remaining=3  generation_remaining=2  → verify_len=2
mixed batch: [6, 2, 1, 0, 2] ✓
```

## Checklist

- [x] Format your code with `pre-commit run --all-files` on the changed files.
- [x] Add unit tests as outlined in the test registration doc.
- [x] Update documentation / docstrings as needed.
- [x] Provide throughput / accuracy verification results as needed.

---
### CI States

Latest PR Test (Base): :x: [Run #31162133949](https://github.com/sgl-project/sglang/actions/runs/31162133949)
Latest PR Test (Extra): :x: [Run #31162132915](https://github.com/sgl-project/sglang/actions/runs/31162132915)

### CODE REVIEW

**Comment by haimianxing** [comment]
Hi maintainers — the lint findings have been addressed in commit 99d6710b: Black formatting is applied, and the two CPU regression files are registered with `register_cpu_ci(..., suite="base-a-test-cpu")` with executable test entry points. Local validation passes (registered-test validator, Black check, and 13 CPU regression tests). Could you please add the `run-ci` label to trigger the required CI suite? Thank you!

**Comment by haimianxing** [comment]
@ch-wan @Qiaolin-Yu Could you add the `run-ci` label to kick off the full test suite? This is a CPU-safe fix for #33454 — the two new test files are CPU-only and pass locally. Thank you!
